sync/src/synchronizer/block_fetcher.rs

```diff
@@ -172,6 +173,25 @@ impl BlockFetcher {
 
         // The headers in `fetch` may be unordered. Sort them by number.
         fetch.sort_by_key(|header| header.number());
-        Some(fetch.iter().map(core::HeaderView::hash).collect())
+
+        let tip = self.active_chain.tip_number();
+        if fetch
+            .last()
+            .map(|header| header.number().saturating_sub(512) > tip)
+            .unwrap_or(false)
+        {
+            if let Some(entry) = inflight.trace_number.get_mut(&(tip + 1)) {
```

Comment:
We don't need to remember the `trace_number` to find the stuck block. Just use the first inflight block in the window. See corresponding comment above.